scripts/debug_tools/reduce.py

```diff
@@ -0,0 +1,304 @@
+import argparse
+import json
+import operator
+import os
+import re
+import shutil
+import stat
+import subprocess
+import sys
+import tempfile
+
+
+def check_files_existence(report_dir):
+    '''
+    This function checks the existence of some files which are required for the
+    reduction process.
+    '''
+    def check(path, isfile):
+        if not (os.path.isfile(path) if isfile else os.path.isdir(path)):
+            print('Error: ' + path + " doesn't exist.")
+            sys.exit(1)
+
+    check(report_dir, False)
+    check(os.path.join(report_dir, 'metadata.json'), True)
+    check(os.path.join(report_dir, 'compiler_includes.json'), True)
+    check(os.path.join(report_dir, 'compiler_target.json'), True)
+    check(os.path.join(report_dir, 'compile_cmd.json'), True)
+
+
+def check_commands_existence():
+    '''
+    This function checks the existence of some commands which are required for
+    the reduction process. A dictionary is returned where the key is the
+    command name and the value is its absolute path.
+    '''
+    def get_path(cmd):
+        proc = subprocess.Popen(['which', cmd], stdout=subprocess.PIPE)
+        return proc.communicate()[0]
+
+    commands = ['CodeChecker', 'creduce', 'clang', 'clang-tidy']
+
+    result = dict()
+
+    for command in commands:
+        path = get_path(command).strip()
+
+        if not path:
+            print('Error: ' + command + ' command not found.')
+            sys.exit(1)
+
+        result[command] = path
+
+    return result
+
+
+def checker_categories():
+    '''
+    This function returns a set of main checker categories i.e. the first part
+    of each checker name.
+    '''
+    checkers_proc = subprocess.Popen(['CodeChecker', 'checkers'],
+                                     stdout=subprocess.PIPE)
+
+    checkers, _ = checkers_proc.communicate()
+    return set(checker[:checker.replace('.', '-').find('-')]
+               for checker in checkers.split())
+
+
+def get_prepared_analyze_command(report_dir,
+                                 workspace_dir,
+                                 codechecker,
+                                 specific_checker=None):
+    '''
+    This function returns a CodeChecker analyze command based on the original
+    one which is stored in the metadata.json file in the report directory.
+
+    report_dir -- The report directory path which is used for the original
+        analysis.
+    workspace_dir -- The workspace directory which contains the new report
+        directory which is used for reduction. The new report directory should
+        be a temporary location because its content is rewritten between the
+        reduction steps.
+    codechecker -- Full path of CodeChecker.
+    specific_checker -- For fastening reduction only this checker will be
+        enabled and all others will be disabled
+    '''
+    def remove_flag(cmd, flag):
```

Comment:
We remove the parameter of the flag too. Maybe `remove_parameterized_flag` function name would be much better or create and extra argument for this function that allows the user to remove only the flag.